waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -239,74 +247,95 @@ public DatabaseMapping databaseMapping(@NotNull String databaseName) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
-              }
-            }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
-          }
+        Map<DatabaseMapping, String> databaseMappingsForPattern = databaseMappingsByDbPattern(databasePattern);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(databaseMappingsForPattern.size());
+        List<Future<List<?>>> futures = new ArrayList<>();
+        String errorMessage = "Can't fetch databases by pattern: {}";
+
+        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsForPattern.entrySet()) {
+          GetAllDatabasesByPatternRequest databasesByPatternRequest = new GetAllDatabasesByPatternRequest(
+              mappingWithPattern.getKey(), mappingWithPattern.getValue(), mappedDbByPrefix, Collections.emptyMap(),
+              null,
+              PREFIXED_RESOLUTION_TYPE);
+          futures.add(executorService.submit(databasesByPatternRequest));
+        }
+
+        Iterator<DatabaseMapping> iterator = databaseMappingsForPattern.keySet().iterator();
+        try {
+          DatabaseMappingUtils.getDatabasesFromFuture(futures, iterator, combined, errorMessage, LOG);
+        } finally {
+          DatabaseMappingUtils.shutdownExecutorService(executorService);
         }
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases() {
         List<String> combined = new ArrayList<>();
-        for (DatabaseMapping mapping : databaseMappings()) {
-          try {
-            List<String> databases = mapping.getClient().get_all_databases();
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
-              }
-            }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases: {}", e.getMessage());
-          }
+        List<DatabaseMapping> databaseMappings = databaseMappings();
+        ExecutorService executorService = Executors.newFixedThreadPool(databaseMappings.size());
+        List<Future<List<?>>> futures = new ArrayList<>();
+        String errorMessage = "Can't fetch databases: {}";
+
+        for (DatabaseMapping mapping : databaseMappings) {
+          GetAllDatabasesRequest allDatabasesRequest = new GetAllDatabasesRequest(mapping, mappedDbByPrefix);
+          futures.add(executorService.submit(allDatabasesRequest));
+        }
+
+        Iterator<DatabaseMapping> iterator = databaseMappings.iterator();
+        try {
+          DatabaseMappingUtils.getDatabasesFromFuture(futures, iterator, combined, errorMessage, LOG);
+        } finally {
+          DatabaseMappingUtils.shutdownExecutorService(executorService);
```

Comment:
OK, so we do call this multiple times, but I think all in this class? So just make it a method here instead of on the Utils class.

Reply:
This happens for both PrefixBasedDatabaseMappingService and StaticDatabaseMappingService, since they both use an executor service that will have to be shutdown :)